extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -187,362 +191,468 @@ oppia.directive('graphViz', function() {
-          $scope.graph.vertices[$scope.state.selectedVertex].label = label;
+          if (angular.isDefined(weight) && angular.isNumber(weight)) {
+            $scope.graph.edges[$scope.state.selectedEdge].weight = weight;
+          }
+          return $scope.graph.edges[$scope.state.selectedEdge].weight;
+        };
+
+
+        // Styling functions
+        var DELETE_COLOR = 'red';
+        var HOVER_COLOR = 'aqua';
+        var SELECT_COLOR = 'orange';
+        var DEFAULT_COLOR = 'black';
+        $scope.getEdgeColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredEdge &&
+              $scope.canDeleteEdge) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.hoveredEdge) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedEdge === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getVertexColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredVertex &&
+              $scope.canDeleteVertex) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.currentlyDraggedVertex) {
+            return HOVER_COLOR;
+          } else if (index === $scope.state.hoveredVertex) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedVertex === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getDirectedEdgeArrowPoints = function(index) {
+          return graphDetailService.getDirectedEdgeArrowPoints(
+            $scope.graph, index);
+        };
+        $scope.getEdgeCentre = function(index) {
+          return graphDetailService.getEdgeCentre($scope.graph, index);
+        };
+      }]
+  };
+});
+
+oppia.factory('graphInputRulesService', [function() {
+  var isIsomophic = function(graph1, graph2) {
+    if (graph1.vertices.length != graph2.vertices.length) {
+      return false;
+    }
+
+    var constructAdjacencyMatrix = function(graph) {
+      var adjMatrix = [];
+      graph.vertices.map(function() {
+        adjMatrix.push([]);
+      });
+      graph.edges.map(function(edge) {
+        var weight = graph.isWeighted ? edge.weight : 1;
+        adjMatrix[edge.src][edge.dst] = weight;
+        if (!graph.isDirected) {
+          adjMatrix[edge.dst][edge.src] = weight;
         }
-        return $scope.graph.vertices[$scope.state.selectedVertex].label;
-      };
-      $scope.selectedEdgeWeight = function(weight) {
-        if ($scope.state.selectedEdge === null) {
-          return '';
+      });
+      return adjMatrix;
+    };
+
+    var adj1 = constructAdjacencyMatrix(graph1);
+    var adj2 = constructAdjacencyMatrix(graph2);
+
+    var nextPermutation = function(permutation) {
```

Comment:
Do you mean a utils factory shared across all files? I'm not very sure about that, since this function won't be used in many places.

What do you think of creating a GraphUtils library to put this and also other parts of the graph rule evaluation helper functions?